// models/dbSchema.js
const mongoose = require('mongoose');
const Schema = mongoose.Schema;
const User = require("./users");

const senderSchema = new mongoose.Schema({
	_id: Schema.Types.ObjectId,
	firstName: {
		type: String
	},
	middleName: String,
	lastName: {
		type: String
	},
	address: String
});

const receiverSchema = new mongoose.Schema({
	_id: Schema.Types.ObjectId,
	firstName: {
		type: String
	},
	middleName: String,
	lastName: {
		type: String
	},
	address: String
});

const statusSchema = new mongoose.Schema({
	_id: Schema.Types.ObjectId,
	stage: {
		type: String,
		default: "NZ Port"
	},
	arrivalDate: {
		type: Date
	},
	estDeliveryDate: {
		type: Date
	},
	actDeliveryDate: {
		type: Date
	},
	receivedBy: {
		type: String
	}
});

const containerLineSchema = new mongoose.Schema({
    _id: Schema.Types.ObjectId,
	count: Number, 
	batchNo: Number,
	trackingNo: {
        type: String
	},
	area: {
        type: String,
		default: ""
	},
	status: statusSchema,
	comment: {
        type: String,
		default: ""
	},
	sender : {
    	type: senderSchema,
		required: true
	},
	receiver : {
    	type: receiverSchema,
		required: true
	}
});

const containerSchema = new mongoose.Schema({
	_id: Schema.Types.ObjectId,
	dateAdded: Date,
	dateLastAccessed: Date,
	containerLine: [containerLineSchema],
	localAccess: {
		type: Schema.Types.ObjectId,
		ref: "User"
	},
	overseasAccess: {
		type: Schema.Types.ObjectId,
		ref: "User"
	}
});

module.exports = mongoose.model("Container", containerSchema);
